packages/errors/src/error/server.ts

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  parseServerResponseErrorBody,
+  ServerResponseErrorBody,
+} from '../payload';
+
+/**
+ * An error thrown as the result of a failed server request.
+ *
+ * The server is expected to respond on the ServerResponseErrorBody format.
+ */
+export class ServerResponseError extends Error {
```

Comment:
Thoughts on making this mirror the response error more closely? Alternatively constructing an error object from the response and setting it as the reason?

I'm thinking version makes it a bit too tricky to have logic based on the error that happened. Let's say we for example want to show a not found page for a 404, but otherwise display an error message, I'm looking to be able to do something like one of these:

```ts
if (error.name === 'NotFoundError') {} // Copy the name or infer from status code

if (error.reason.name === 'NotFoundError') {} // Use a reason property, with error being a ServerReponseError

if (hasReason(error, 'NotFoundError') {} // Perhaps a utility around it, especially if supporting multiple reasons

if (error.status === 404) {} // Skip the name, just check the code
```

I think especially for the service-to-service communication it's better to be a bit more symmetric in the error forwarding, where you essentially are able to access the same error structure in the calling service as the error that was thrown upstream, for example though `responseError.reason`.

Reply:
Hm sounds good but more precisely in what way would we make it mirror the response more closely? It's passed basically verbatim in the error.

Reply:
Preserving name, stack, message, any extra fields. As a proper error object too that is, as opposed to the plain body

Reply:
Hm but CAN you assign the stack and stuff in an Error in a meaningful way? Just had a thing about that https://github.com/backstage/backstage/pull/4889#discussion_r591468730

Reply:
So I guess we could essentially json stringify the original Error on the server side, absolutely, to preserve as much as possible about it. But recreating it on the receiving end? Not as sure about that one. Maybe? Well have to look into it

Reply:
Done! It now actually serializes and deserializes the cause formally, and returns a wrapping error type that holds the envelope info plus the cause